### DISCUSSION TITLE
sidebar/sanfene/javathread

### DISCUSSION THREAD

**Discussion by giscus** [itwanger/tobebetterjavaer-giscus#84] [Announcements]
# sidebar/sanfene/javathread

图文详解 60 道Java并发面试高频题，这次面试，一定吊打面试官

https://tobebetterjavaer.com/sidebar/sanfene/javathread.html

**Comment by zdcat** [comment]
感觉重锁的例子等待室举例的不太好，等待室不如叫候诊室，也就是EntryList，没有竞争到锁的就去候诊室，去等待室应该是缺器材（对应根据业务逻辑去wait）的情况。

**Comment by spark2025** [comment]
老哥，第二个问题中有一句话：线程是 CPU 分配资源的基本单位。上学的时候记得书上讲的是 进程资源分配的基本单位，线程是调度的基本单位。脑子现在有点不会了

**Comment by 756434644** [comment]
图片内容挂了

**Comment by xiemeng9462** [comment]
已阅

**Comment by Kddk-K** [comment]
9.28一刷

**Comment by TheFivePercent** [comment]
wait是非静态方法吧，应该对象.wait()来调用，不能单单是wait()

**Comment by szy0417-maker** [comment]
2025.01.13 一刷

**Comment by hachinekooo** [comment]
1.17

**Comment by Slinet6056** [comment]
线程池工作流程举例的第三种情况描述 `排队区的人立马就冲了过去` 是不是不太准确，应该是让新来的人直接去新窗口，排队的人还得继续排

**Comment by EVAKKKK** [comment]
Entry 继承了 WeakReference，它限定了 key 是一个弱引用，弱引用的好处是当内存不足时，JVM 会回收 ThreadLocal 对象，并且将其对应的 Entry.value 设置为 null，这样可以在很大程度上避免内存泄漏。弱引用不管内存够不够都会回收

**Comment by shenlanoo7** [comment]
纯抄

**Comment by Xiecanquan** [comment]
AQS的CLH队列，线程获取锁失败后进入队尾应该是挂起而不是自旋？是吗各位大佬，ai说的我也不太确定

**Reply by Tinkol1998** [reply]
是的 public final void acquire(int arg) {
    if (!tryAcquire(arg) &&
        acquireQueued(addWaiter(Node.EXCLUSIVE), arg))
        selfInterrupt();
}

**Reply by Tinkol1998** [reply]
先自旋不成功就加入CLH

**Comment by Xiecanquan** [comment]
ThreadLocal内存泄漏那部分，“找到后会调用 Entry 的 clear 方法，将 Entry 的 value 设置为 null。”这部分是不是写错了，好像是Entry的clear方法不会将Entry的value置为null，只是将Entry的value置为null，然后调用expungeStaleEntry方法将value置为null，并对key为null的Entry进行清除，问ai总结出来的，对吗各位大佬

**Reply by Xiecanquan** [reply]
我上面说错了，是Entry的clear方法只是会对Entry的key置为null，而不能将Entry的value置为null，将value置为null是expungeStaleEntry方法的工作
